### PR TITLE
[util] Add ray/util/counter.h to clean up metrics tracking code

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds a Counter util class, split out from https://github.com/ray-project/ray/compare/master...ericl:ray:task-metric-3?expand=1

It is a simple wrapper around absl::flat_hash_map that handles entry cleanup.

### CODE REVIEW

File: src/ray/util/counter.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+#pragma once
+
+#include <list>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/util/logging.h"
+
+/// \class Counter
+///
+/// This container implements counter behavior on top of an absl hash table. Counter
```
Comment (rkooo567) [inline src/ray/util/counter.h]: Should we mention this class is not thread-safe?

File: src/ray/util/counter_map.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+/// \class CounterMap
+///
+/// This container implements counter behavior on top of an absl hash table. CounterMap
+/// entries will be automatically cleaned up when they fall back to zero. CounterMap
+/// entries are not allowed to be negative. A callback can be set to run when any
+/// counter entry changes.
+///
+/// For example, this can be used to track the number of running tasks broken down
+/// by their function name, or track the number of tasks by (name, state) pairs.
+template <typename K>
```
Comment (rkooo567) [inline src/ray/util/counter_map.h:31]: Duplicated file?

**Review by rkooo567 — APPROVED** [review]
Seems like there are 2 duplicated files?

**Review by ericl — COMMENTED** [review]
Fixed.

**Pull request merged**